kv/kvserver: TestPromoteNonVoterInAddVoter failed

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://tanzanite.cluster.engflow.com/invocations/default/ba0f3dab-c457-49f3-9256-6076ef741a8d?testReportRun=5&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.1 @ [ac2b6541807c29116911dd08b96538935477667d](https://github.com/cockroachdb/cockroach/commits/ac2b6541807c29116911dd08b96538935477667d):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestPromoteNonVoterInAddVoter1516441687
    test_log_scope.go:76: use -show-logs to present logs inline
    replicate_queue_test.go:2330: 
        	Error Trace:	pkg/kv/kvserver/replicate_queue_test.go:2330
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 5
        	            	Add voter events: [{r75:/{Table/104-Max} [(n7,s7):1, (n2,s2):7, (n1,s1):3, (n3,s3):4, (n6,s6):5, (n4,s4):6, (n5,s5):8LEARNER, next=9, gen=41] <nil> <nil> (n5,s5):8LEARNER <nil> range under-replicated {"Target":"s5, diversity:0.92, converges:0, balance:0, rangeCount:74"}} {r75:/{Table/104-Max} [(n7,s7):1, (n2,s2):7, (n1,s1):3, (n3,s3):4, (n6,s6):5, (n4,s4):6, (n5,s5):8, next=9, gen=42] <nil> <nil> (n5,s5):8 <nil> range under-replicated {"Target":"s5, diversity:0.92, converges:0, balance:0, rangeCount:74"}}]
    exectrace.go:39: execution trace written to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvserver/kvserver_test/shard_13_of_48_run_5_of_10/test.outputs/logTestPromoteNonVoterInAddVoter1516441687/exectrace_goid_60579.bin
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestPromoteNonVoterInAddVoter1516441687
--- FAIL: TestPromoteNonVoterInAddVoter (40.35s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139318 kv/kvserver: TestPromoteNonVoterInAddVoter failed [C-bug C-test-failure O-robot P-3 T-kv branch-master]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48597

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -2196,18 +2196,18 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 		if err := forceScanOnAllReplicationQueues(tc); err != nil {
 			return err
 		}
-		rangeCount := -1
-		allEqualRangeCount := true
-		iterateOverAllStores(t, tc, func(s *kvserver.Store) error {
-			if rangeCount == -1 {
-				rangeCount = s.ReplicaCount()
-			} else if rangeCount != s.ReplicaCount() {
-				allEqualRangeCount = false
-			}
-			return nil
-		})
-		if !allEqualRangeCount {
-			return errors.New("Range counts are not all equal")
+		s, err := sqlutils.RowsToDataDrivenOutput(sqlutils.MakeSQLRunner(tc.Conns[0]).Query(t, `
+SELECT * FROM (
+    SELECT
+        range_id,
+        array_length(voting_replicas, 1) AS vc,
+        COALESCE(array_length(non_voting_replicas, 1), 0) AS nvc
+    FROM crdb_internal.ranges_no_leases
+) WHERE vc != 7 OR nvc > 0 ORDER BY range_id ASC LIMIT 1
+`))
+		require.NoError(t, err)
+		if len(s) > 0 {
+			return errors.Errorf("still upreplicating:\n%s", s)
 		}
 		return nil
 	})

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -2126,17 +2126,17 @@ func iterateOverAllStores(
 //
 // ZONE survival configuration:
 // Region 1: 3 of [n1 (voter) n2 (voter) n3 (voter)]
-// Region 2: 1 of [n4 (non-voter) n5 (non-voter)]
-// Region 3: 1 of [n6 (non-voter) n7 (non-voter)]
+// Region 2: 1 of [n4 or n5 (non-voter)]
+// Region 3: 1 of [n6 or n7 (non-voter)]
 // to REGION survival configuration:
-// Region 1: 2 of [n1 (voter) n2 (voter) n3 (voter)]
-// Region 2: 2 of [n4 (voter) n5 (voter)]
-// Region 3: 1 of [n6 (voter) n7 (voter)]
+// Region 1: 2 of [two of n1-n3 (voter)]
+// Region 2: 2 of [n4 (voter) and n5 (voter)]
+// Region 3: 1 of [n6 or n7 (voter)]
 //
 // Here we have 7 stores: 3 in Region 1, 2 in Region 2, and 2 in Region 3.
 //
-// The expected behaviour is that there should not be any add voter events in
-// the range log where the added replica type is a LEARNER.
+// What the test wants to see is that when we switch from ZONE to REGION,
+// the non-voters in Region 2 and Region 3 are promoted to voters.
 func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	scope := log.Scope(t)
@@ -2216,6 +2216,7 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	_, err := db.Exec("CREATE TABLE t (i INT PRIMARY KEY, s STRING)")
 	require.NoError(t, err)
 
+	log.Infof(ctx, "test setting ZONE survival configuration")
 	// ZONE survival configuration.
 	setConstraintFn("TABLE t", 5, 3,
 		", constraints = '{\"+region=2\": 1, \"+region=3\": 1}', voter_constraints = '{\"+region=1\": 3}'")
@@ -2260,6 +2261,11 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	})
 
 	// REGION survival configuration.
+	log.Infof(ctx, "test setting REGION survival configuration")
+	// Clear the rangelog so that we can rest assured to only pick up events
+	// resulting from the zone config change.
+	_, err = tc.Conns[0].ExecContext(ctx, `DELETE FROM system.rangelog WHERE TRUE`)
+	require.NoError(t, err)
 	setConstraintFn("TABLE t", 5, 5,
 		", constraints = '{}', voter_constraints = '{\"+region=1\": 2, \"+region=2\": 2, \"+region=3\": 1}'")
 	require.NoError(t, err)

